### PR TITLE
docs: 25702 Design: validate block stream equivalence via block-stream → PCES replay

### LINKED ISSUE #25702: Design: validate block stream equivalence via block-stream → PCES replay
**Description:**

We want to verify that a consensus node running the current software, when fed historical production traffic, deterministically reproduces the same block stream (block hashes and state hashes) as the original production run. This is an end-to-end regression test for the consensus → execution → block-production pipeline against real-world data.

The proposed approach: reconstruct gossip events from production block stream files, write them as PCES files, and replay them on a single node started from a production state snapshot. PCES replay runs the full pipeline before gossip starts, so one node produces all blocks with no peer communication, and the output can be compared by hash. Determinism holds because the hashgraph derives consensus from the event DAG, and block-stream order is a valid topological order.

Produce a design doc covering:

- The overall test goal, approach, and why a single node suffices and is deterministic.
- The block stream → PCES conversion mechanism (reconstructing `GossipEvent`s from `EventHeader` + `signed_transaction` items, resolving parents, recomputing event hashes, handling the unsigned-event case).
- The components required: the conversion tool, the unsigned-event intake path, block-signing behavior during replay (mock vs. real TSS), and the inputs/environment (production roster, this node's private key, state snapshot).
- The test flow and the hash-based validation strategy.
- Open questions, edge cases (e.g. trailing unsigned blocks), and relevant configuration.

**Outcome:** a reviewed design doc in the repo that scopes the implementation work — primarily the conversion tool and the unsigned-event intake path.

### PR DESCRIPTION
**Pull request by imalygin**
**Description**:

Adds a design doc for an end-to-end test that validates block stream equivalence — verifying that a node on the current software, replaying historical production traffic, deterministically reproduces the same block stream (block and state hashes) as the original run.

The doc covers the overall approach (reconstruct events from block files, replay them as PCES on a single node that mirrors the full production configuration, before gossip starts), the block-stream → PCES conversion mechanism, the components required, the test flow and hash-based validation, and the open questions and edge cases.

Doc only — no code changes. Looking for review of the approach and the scoping of follow-up implementation work.

**Related issue(s)**:

Fixes #25702 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+runs the complete `intake → hashgraph → transaction handling → block production`
+pipeline *before* gossip starts, so this one node produces all blocks with no
+peer communication. The output is compared against the original blocks/state by
+hash.
+
+Why this works:
+
+- **Determinism.** The hashgraph derives rounds, consensus timestamps, and
+  consensus order from the event DAG, not from arrival order. Block-stream order
+  is a valid topological order, so the reconstructed DAG yields identical
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: Just a clarification here. The only reason we need to feed the events into replay in topological order is because of memory constraints and how we chose to design PCES. In a theoretical sense, the order or events replayed does not matter at all - the output of the consensus algorithm will be deterministic because the same graph of events is provided as input (order does not matter). Perhaps this clarification is not required for this project, but if AI will be consuming this document it might help. The current implication from the text is that each node must receive events in topological order which is not how gossip works. There are components to put events into topological order before events hit the consensus algorithm (the orphan buffer).

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+## Components
+
+### 1. Block stream → PCES conversion tool
+Reads `.blk.gz` files, reconstructs `GossipEvent` records, and writes PCES files.
+Detailed in the next section. Implemented as an enhancement to the existing PCES
+tooling so the normal `PcesFileTracker` replay path can consume its output
+unchanged.
+
+### 2. Unsigned-event intake path
+Reconstructed events lack the creator's `GossipEvent.signature`. PCES-sourced
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: I forgot about the fact that events in the block stream do not have signatures. This is a problem and will require code changes for replay to work. Using the flag mentioned will not work - that only avoids ISSes, and is not required for this scenario.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+## Block Stream → PCES Conversion Mechanism
+
+### What the block stream provides per event
+- `EventHeader` block item → `EventCore` (creator, birth round, time created,
+  version) and parent references (`ParentEventReference`: a full `EventDescriptor`
+  for out-of-block parents, or an in-block `index`).
+- The `signed_transaction` items following an `EventHeader`, up to the next
+  header → that event's transactions, including system event transactions such
+  as state signatures.
+- `RedactedItem.signed_transaction_hash` → the hash of any redacted event
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md:188]: Does the handling logic know what to do with a redacted item? What about the state change for a redacted transaction?

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+   `signed_transaction_hash`.
+3. **Resolve parents.** Use `event_descriptor` references directly; resolve
+   `index` references to the in-block event by position. Build each
+   `EventDescriptor` as `(hash, creatorId, birthRound)`, where `hash` is the
+   reconstructed parent event hash.
+4. **Compute the event hash** with the same algorithm as `PbjStreamHasher`:
+   hash `EventCore` + parent `EventDescriptor`s + double-hashed transactions.
+   Process events in topological order so parent hashes are available before
+   their children. The hash excludes the signature.
+5. **Assemble the `GossipEvent`**: `event_core` + `parents` + `transactions`,
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: FYI, all steps up to and include step 5 already have code written. Check out the `EventHashBlockStreamValidator` and `BlockStreamEventBuilder` classes.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+   with the `signature` field left empty/placeholder.
+6. **Write PCES files**: a 4-byte version header (`2` = `PROTOBUF_EVENTS`)
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: I think you'll want to write the PCES files using the `CommonPcesWriter` class.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+   followed by length-delimited `GossipEvent` records, emitted in block-stream
+   order. Use the standard PCES filename convention (sanitized timestamp,
+   sequence number, min/max birth round, origin). Set `origin` consistent with
+   the snapshot's starting round and birth-round bounds to cover the events.
+   Reuse `PcesFile` / `PcesMutableFile` for descriptor creation and writing.
+
+### Notes
+- **Byte-exact hashes are achievable.** `EventCore` and parent descriptors come
+  straight from the stream, transaction bytes are the `signed_transaction`
+  items, and the double-hash recovers redacted transactions from their hashes.
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: The production event hasher does not know how to handle redacted transactions when computing the event hash. Additionally, the `GossipEvent` does not allow for redacted transactions, unless we want to get fancy and somehow build into the transaction bytes an indicator that this is only a hash, and add code to look for this on every transaction with performing block-stream-event-replay only.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+- Block proofs may be **valid but not byte-identical** to production, because
+  threshold-signature aggregation is not bit-stable. Under Tier 1
+  (`forceMockSignatures`) the proof is a deterministic mock signature, not a real
+  TSS signature. Compare block and state hashes, not raw proof bytes.
+
+## Open Questions / Risks
+- **Trailing blocks of the window.** Block proofs depend on partial-signature
+  transactions that appear in later blocks; the last few blocks of the replay
+  window may remain pending/unsigned. Compare up to the last fully-signable
+  block, or extend the window past the comparison target.
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: Just to clarify, this must mean to extend the window of blocks for which events are extracted from past the comparison target. This should include events with the required signatures.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+## Relevant Configuration
+- `tss.forceMockSignatures=true` — Tier 1: always-ready signer producing
+  deterministic mock block signatures, enabling single-node block production
+  without TSS. Omit (with hinTS enabled) for Tier 2 real-TSS validation.
+- `tss.useDeterministicHintsSignatures=true` — Tier 2: incorporate partial
+  signatures in consensus order so block proofs are byte-reproducible.
+- `event.preconsensus.limitReplayFrequency=false` — replay as fast as the
+  pipeline allows (removes the default 5000 events/s cap).
+- `event.preconsensus.replayHealthThreshold` — tune backpressure for very long
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: This configuration should not be changed. A very long replay window does not require changing this value. It changes the minimum duration of node unhealthiness before pausing the event replay in order to allow the node to recover.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  signing in both tiers needs none). It is described as planned but not yet
+  implemented.
+
+## Relevant Configuration
+- `tss.forceMockSignatures=true` — Tier 1: always-ready signer producing
+  deterministic mock block signatures, enabling single-node block production
+  without TSS. Omit (with hinTS enabled) for Tier 2 real-TSS validation.
+- `tss.useDeterministicHintsSignatures=true` — Tier 2: incorporate partial
+  signatures in consensus order so block proofs are byte-reproducible.
+- `event.preconsensus.limitReplayFrequency=false` — replay as fast as the
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: I do not recommend changing this value unless there is a problem during replay or it is too slow.

File: platform-sdk/docs/proposals/event-replay/event-replay.md (hunk 1 above)
Comment (imalygin) [reply platform-sdk/docs/proposals/event-replay/event-replay.md]: sounds good, I'll incorporate your comment into the doc.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+window past the comparison target.
+
+### 4. Inputs and environment
+
+- A production signed-state snapshot at the round immediately preceding the
+  first reconstructed block.
+- Block stream files covering the target window, contiguous from that round.
+- The full production roster (all members' public keys), with the node running in
+  its normal multi-node production configuration — **not** single-node mode. The
+  other roster members are simply not started.
```
Comment (jsync-swirlds) [inline platform-sdk/docs/proposals/event-replay/event-replay.md:166]: Note, we'd also need a blocking netfilter setup on the node running the test to ensure that any attempt to connect to the IPs for other nodes goes nowhere, because the other hosts _might_ actually be running in live production environments, so it's important to block attempts to communicate with them.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+  order is a practical requirement of how PCES replay works and the bounded
+  memory of the orphan buffer — *not* a consensus requirement. (In normal
+  operation gossip delivers events out of order; the orphan buffer holds each
+  event until its parents arrive and releases them in topological order before
+  the hashgraph sees them.) Block-stream order is already a valid topological
+  order, so it keeps orphan-buffer pressure low and needs no re-sorting.
+- **Single node, production config.** The node mirrors the production setup
+  (full roster, normal multi-node configuration) and is *not* run in single-node
+  mode — the other roster members are simply absent. Consensus never actually
+  starts: PCES replay completes before `platformCoordinator.startGossip()`, so
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md]: It's not accurate to say that consensus never actually starts, but it is accurate to say that consensus will not proceed past the point it got to on the original network because none of the other nodes required to do so are present.

**Review by jsync-swirlds — APPROVED** [review]
This looks good to me.
One comment about the testing environment to prevent unintended side effects; but that is just an item to note.

File: platform-sdk/docs/proposals/event-replay/event-replay.md (hunk 10 above)
Comment (imalygin) [reply platform-sdk/docs/proposals/event-replay/event-replay.md:166]: Fair point, I'll update the doc.

File: platform-sdk/docs/proposals/event-replay/event-replay.md [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+  transaction, sufficient to recompute the event hash.
+
+The block stream is designed to support event reconstruction: all transactions
+in an event appear in the stream, and the per-transaction double-hash in
+`PbjStreamHasher` lets a redacted transaction still contribute its hash to the
+event hash.
+
+### Reconstruction (steps 1–5 already implemented)
+
+Steps 1–5 below are implemented today in `BlockStreamEventBuilder` (in
```
Comment (poulok) [inline platform-sdk/docs/proposals/event-replay/event-replay.md:198]: We should explain why the absence of stale events in the block stream is not a problem:

Stale events never reach consensus and are therefore never included in the block stream. They can, however, have child events that do reach consensus and are present in the block stream. That means there will be events in the block stream that reference a parent that is not in the block stream. This does not cause any problem because the child events will sit in the orphan buffer until the stale parent becomes ancient, then it will continue on. Stale events do not impact the result of consensus, so the output will be the same.

**Review by poulok — APPROVED** [review]
